srai/regionizers/_spherical_voronoi.py

```diff
@@ -0,0 +1,252 @@
+
+This module contains spherical voronoi implementation based on
+SphericalVoronoi function from scipy library.
+"""
+
+from math import ceil, sqrt
+from typing import List, Tuple
+
+import numpy as np
+import numpy.typing as npt
+from haversine import haversine
+from pymap3d import Ellipsoid, ecef2geodetic, geodetic2ecef
+from scipy.spatial import SphericalVoronoi, geometric_slerp
+from shapely.geometry import MultiPolygon, Point, Polygon, box
+from spherical_geometry.polygon import SphericalPolygon
+
+# LON: 0; LAT: 0
+POINT_FRONT = (1.0, 0.0, 0.0)
+# LON: 180; LAT: 0
+POINT_BACK = (-1.0, 0.0, 0.0)
+# LON: 0; LAT: 90
+POINT_TOP = (0.0, 0.0, 1.0)
+# LON: 0; LAT: -90
+POINT_BOTTOM = (0.0, 0.0, -1.0)
+# LON: -90; LAT: 0
+POINT_LEFT = (0.0, -1.0, 0.0)
+# LON: 90; LAT: 0
+POINT_RIGHT = (0.0, 1.0, 0.0)
+
+SPHERE_PARTS = [
+    SphericalPolygon([POINT_FRONT, POINT_TOP, POINT_RIGHT, POINT_FRONT]),
+    SphericalPolygon([POINT_RIGHT, POINT_TOP, POINT_BACK, POINT_RIGHT]),
+    SphericalPolygon([POINT_BACK, POINT_TOP, POINT_LEFT, POINT_BACK]),
+    SphericalPolygon([POINT_LEFT, POINT_TOP, POINT_FRONT, POINT_LEFT]),
+    SphericalPolygon([POINT_FRONT, POINT_RIGHT, POINT_BOTTOM, POINT_FRONT]),
+    SphericalPolygon([POINT_RIGHT, POINT_BACK, POINT_BOTTOM, POINT_RIGHT]),
+    SphericalPolygon([POINT_BACK, POINT_LEFT, POINT_BOTTOM, POINT_BACK]),
+    SphericalPolygon([POINT_LEFT, POINT_FRONT, POINT_BOTTOM, POINT_LEFT]),
+]
+SPHERE_PARTS_BOUNDING_BOXES = [
+    box(minx=0, miny=0, maxx=90, maxy=90),
+    box(minx=90, miny=0, maxx=180, maxy=90),
+    box(minx=-180, miny=0, maxx=-90, maxy=90),
+    box(minx=-90, miny=0, maxx=0, maxy=90),
+    box(minx=0, miny=-90, maxx=90, maxy=0),
+    box(minx=90, miny=-90, maxx=180, maxy=0),
+    box(minx=-180, miny=-90, maxx=-90, maxy=0),
+    box(minx=-90, miny=-90, maxx=0, maxy=0),
+]
+
+
+class SphereEllipsoid(Ellipsoid):  # type: ignore
+    """A sphere ellipsoid."""
+
+    def __init__(self) -> None:
+        """
+        Sphere ellipsoid extends Ellipsoid from pymap3d [1] library.
+
+        Class is used for mapping lat/lon coordinates
+        from and to cartesian x/y/z values on a unit sphere.
+        Required for spherical voronoi algorithm.
+
+        References:
+            [1] https://github.com/geospace-code/pymap3d
+        """
+        self.semimajor_axis = 1
+        self.semiminor_axis = 1
+        self.flattening = (self.semimajor_axis - self.semiminor_axis) / self.semimajor_axis
+        self.thirdflattening = (self.semimajor_axis - self.semiminor_axis) / (
+            self.semimajor_axis + self.semiminor_axis
+        )
+        self.eccentricity = sqrt(2 * self.flattening - self.flattening**2)
+
+
+def map_to_geocentric(lon: float, lat: float, ell: Ellipsoid) -> Tuple[float, float, float]:
+    """
+    Wrapper for a geodetic2ecef function from pymap3d library.
+
+    Args:
+        lon (float): longitude of a point in a wgs84 crs.
+        lat (float): latitude of a point in a wgs84 crs.
```

Comment:
I would omit types of variables in docstring as it is against Dry. I think that the API docs will infer it automatically. Please check and fix everywhere

Reply:
Those types are automatically generated by autodocstring extension in vs code. I will have to disable it in future.

Reply:
We will stay with type hinting for brevity.